lib/core/pktmbuf/pktmbuf.h

```diff
@@ -1153,6 +1194,101 @@ const char *cne_get_tx_ol_flag_name(uint64_t mask);
  */
 int cne_get_tx_ol_flag_list(uint64_t mask, char *buf, size_t buflen);
 
+/**
+ * Return the metadata buffer address
+ *
+ * @param m
+ *   The pktmbuf_t structure pointer
+ * @return
+ *   NULL on error or pointer to metadata buffer
+ */
+static inline void *
+pktmbuf_metadata(const pktmbuf_t *m)
+{
+    pktmbuf_info_t *p;
+
+    if (!m)
+        return NULL;
+
+    if (((p = m->pooldata) != NULL) && p->metadata)
+        return CNE_PTR_ADD(p->metadata, (m->meta_index * p->metadata_bufsz));
+    else
+        return CNE_PTR_ADD(m, sizeof(pktmbuf_t)); /* default to metadata in pktmbuf headroom */
+}
+
+/**
+ * Return the size of the metadata buffer if external or headroom size if internal buffer.
+ *
+ * @param m
+ *   The pktmbuf_t structure pointer
+ * @return
+ *   -1 if pktmbuf pointer is NULL or size of metadata buffer or headroom size if internal.
+ */
+static inline uint32_t
+pktmbuf_metadata_bufsz(const pktmbuf_t *m)
+{
+
+    pktmbuf_info_t *p;
+
+    if (!m)
+        return -1;
+
+    if (((p = m->pooldata) != NULL) && p->metadata)
+        return p->metadata_bufsz;
+    else
+        return (uint32_t)pktmbuf_headroom(m);
+}
+
+/**
+ * Construct a pktmbuf_pool_cfg_t structure with the given parameters.
+ *
+ * @param c
+ *   Pointer to the pktmbuf_pool_cfg_t structure
+ * @param addr
+ *   Address of the pktmbuf buffers, must not be NULL
+ * @param bufcnt
+ *   Number of buffers in the pktmbuf pool, must not be zero
+ * @param bufsz
+ *   Size of each buffer in the pktmbuf pool, must not be zero
+ * @param cache_sz
+ *   Cache size of the mempool to use in allocate/free buffers, can be zero
+ * @param metadata
+ *   The pointer of the metadata buffer pool, can be NULL.
+ * @param metadata_bufsz
+ *   The size of each metadata buffer in the pool, if metadata is valid can not be zero
+ *   and needs to be a multiple of a cacheline.
+ * @param ops
+ *   The pktmbuf ops structure pointer, can be NULL.
+ * @return
+ *   0 on success or -1 on error.
+ */
+static inline int
+pktmbuf_pool_cfg(pktmbuf_pool_cfg_t *c, void *addr, uint32_t bufcnt, uint32_t bufsz,
+                 uint32_t cache_sz, void *metadata, uint32_t metadata_bufsz, mbuf_ops_t *ops)
+{
+    if (c) {
```

Comment:
It's a little challenging to immediately follow the nested if statements with all the returns. Would you consider checking all the values for null/zero at the top? I think you can avoid some of hte nested if statements. I think the following more clearly conveys the constraints.

```
if (!c || !addr || !bufcnt || !bufsz)
  return -1;

if ((!metadata && metadata_bufsz) || (metadata && !metadata_bufsz))
  return -1;

if (metadata && ((metadata_bufsz % CNE_CACHE_LINE_SIZE) != 0))
  return -1;

c->addr = addr;
c->bufcnt = bufcnt;
c->bufsz = bufsz;
c->cache_sz = cache_sz;
c->ops = ops;
c->metadata = metadata;
c->metadata_bufsz = metadata_bufsz;

return 0;
```

Also, this doesn't really need to be a static inline function. Would you consider putting this in pktmbuf.c?

Reply:
Your suggestion is much easy to read and understand. I just tried to make it too hard :-(